File: javascript/23-max_depth_binary_tree.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

//    /**
//     * @param {TreeNode} root
//     * @return {number}
//     */

// method 1
var maxDepth = function (root) {
  let depth = 0;

  let bfs = (node, level) => {
    if (node === null) return;
    if (level > depth) {
      depth = level;
    }
    bfs(node.left, level + 1);
    bfs(node.right, level + 1);
  };
  bfs(root, 1);
  return depth;
};

// method 2
var maxDepth2 = function (root) {
  if (root === null) return 0;
  let l = maxDepth(root.left);
  let r = maxDepth(root.right);
  return 1 + Math.max(l, r);
};

File: python/23-max_depth_binary_tree.py


class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if (root == None):
            return 0
        l = self.maxDepth(root.left)
        r = self.maxDepth(root.right)
        return 1 + max(l, r)


class Solution2(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.depth = 0

        def bfs(node, level):
            if (node == None):
                return
            if level > self.depth:
                self.depth = level
            bfs(node.left, level + 1)
            bfs(node.right, level + 1)
        bfs(root, 1)
        return self.depth
